java/ql/src/experimental/Security/CWE/CWE-074/JndiInjectionLib.qll

```diff
@@ -0,0 +1,91 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+import DataFlow
+import experimental.semmle.code.java.frameworks.Jndi
+import experimental.semmle.code.java.frameworks.spring.SpringJndi
+import experimental.semmle.code.java.frameworks.Shiro
+
+/**
+ * A taint-tracking configuration for unvalidated user input that is used in JNDI lookup.
+ */
+class JndiInjectionFlowConfig extends TaintTracking::Configuration {
+  JndiInjectionFlowConfig() { this = "JndiInjectionFlowConfig" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof JndiInjectionSink }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node.getType() instanceof PrimitiveType or node.getType() instanceof BoxedType
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    compositeNameStep(node1, node2)
+  }
+}
+
+/**
+ * JNDI sink for JNDI injection vulnerabilities, i.e. 1st argument to `lookup`, `lookupLink`,
+ * `doLookup`, `rename`, `list` or `listBindings` method from `InitialContext`.
+ */
+predicate jndiSinkMethod(Method m, int index) {
+  m.getDeclaringType().getAnAncestor() instanceof TypeInitialContext and
+  (
+    m.hasName("lookup") or
+    m.hasName("lookupLink") or
+    m.hasName("doLookup") or
+    m.hasName("rename") or
+    m.hasName("list") or
+    m.hasName("listBindings")
+  ) and
+  index = 0
+}
+
```

Comment:
`javax.management.remote.JMXServiceURL` constructor is also a nice JNDI Injection sink

Reply:
I've added the following sinks:
* `JMXConnectorFactory.connect(new JMXServiceURL(tainted))`
* `JMXConnectorFactory.newJMXConnector(new JMXServiceURL(tainted), sth).connect()`
* `new RMIConnector(new JMXServiceURL(tainted), sth).connect()`

I assume that the invocation of the `connect` method is the actual sink, not the construction of `JMXServiceURL`. Let me know if that's correct.

Reply:
Yep, although sometimes is ok to report the issue before the actual sink. For example, a JMXServiceURL may be constructed in user code and then passed to a 3rd party or even custom library that the analyzer has no visibility on. So if connection is established outside user code, the analyzer may miss that issue. That depends on many factors of course, but in this case, I think it would be safe to hoist the sink to the `JMXServiceURL` constructor. No need to change your PR though, its ok, thanks!

Reply:
Sure. It should be very easy to change it in the future, if someone decides to do this, but I'll leave it as it is for now.